cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go

```diff
@@ -0,0 +1,361 @@
+	var jsonStr = []byte(``)
+	packetAuthToken := os.Getenv("PACKET_AUTH_TOKEN")
+	url := "https://api.packet.net/projects/" + mgr.projectID + "/devices"
+	req, err := http.NewRequest("GET", url, bytes.NewBuffer(jsonStr))
+	req.Header.Set("X-Auth-Token", packetAuthToken)
+	req.Header.Set("Content-Type", "application/json")
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+		// klog.Fatalf("Error listing nodes: %v", err)
+	}
+	defer resp.Body.Close()
+
+	fmt.Println("response Status:", resp.Status)
+	body, _ := ioutil.ReadAll(resp.Body)
+	var devices Devices
+	json.Unmarshal([]byte(body), &devices)
+	return &devices
+}
+
+// nodeGroupSize gets the current size of the nodegroup as reported by packet tags.
+func (mgr *packetManagerRest) nodeGroupSize(nodegroup string) (int, error) {
+	devices := mgr.listPacketDevices()
+	// Get the count of devices tagged as nodegroup members
+	count := 0
+	for _, d := range devices.Devices {
+		if Contains(d.Tags, "k8s-cluster-"+mgr.clusterName) && Contains(d.Tags, "k8s-nodepool-"+nodegroup) {
+			count++
+		}
+	}
+	fmt.Println(len(devices.Devices), count)
+	return count, nil
+}
+
+func randString8() string {
+	n := 8
+	rand.Seed(time.Now().UnixNano())
+	letterRunes := []rune("acdefghijklmnopqrstuvwxyz")
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letterRunes[rand.Intn(len(letterRunes))]
+	}
+	return string(b)
+}
+
+// createNodes provisions new nodes on packet and bootstraps them in the cluster.
+func (mgr *packetManagerRest) createNodes(nodegroup string, nodes int) error {
+	klog.Infof("Updating node count to %d for nodegroup %s", nodes, nodegroup)
+	cloudinit, err := base64.StdEncoding.DecodeString(mgr.cloudinit)
+	if err != nil {
+		log.Fatal(err)
+		return fmt.Errorf("Could not decode cloudinit script: %v", err)
+	}
+
+	udvars := CloudInitTemplateData{
+		BootstrapTokenID:     os.Getenv("BOOTSTRAP_TOKEN_ID"),
+		BootstrapTokenSecret: os.Getenv("BOOTSTRAP_TOKEN_SECRET"),
+		APIServerEndpoint:    mgr.apiServerEndpoint,
+	}
+	ud := renderTemplate(string(cloudinit), udvars)
+	hn := "k8s-" + mgr.clusterName + "-" + nodegroup + "-" + randString8()
+
+	reservation := ""
+	if mgr.reservation == "require" || mgr.reservation == "prefer" {
+		reservation = "next-available"
+	}
+
+	cr := DeviceCreateRequest{
+		Hostname:              hn,
+		Facility:              []string{mgr.facility},
+		Plan:                  mgr.plan,
+		OS:                    mgr.os,
+		ProjectID:             mgr.projectID,
+		BillingCycle:          mgr.billing,
+		UserData:              renderTemplate(ud, udvars),
+		Tags:                  []string{"k8s-cluster-" + mgr.clusterName, "k8s-nodepool-" + nodegroup},
+		HardwareReservationID: reservation,
+	}
+
+	resp, err := createDevice(&cr)
+	if err != nil || resp.StatusCode > 299 {
+		// If reservation is preferred but not available, retry provisioning as on-demand
+		if reservation != "" && mgr.reservation == "prefer" {
```

Comment:
Not follow why any error would mean that the reservation is not available? This is not checking that the API failed with an error that means reservation not available, right? This will just retry on any error if the reservation is configured as "prefer", right?

Or am I missing something?

Reply:
Yes, that's right. If the response is not within the 200-299 range (which should mean in generic HTTP semantics that the create device request was successful) then cluster-autoscaler just panics and prints the error before dying, unless a reservation was set as preferred rather than required, in which case it will retry the request once, omitting the hardware_reservation_id param.

The Packet API with a `503 Service Unavailable` response with the following body: `{"errors":["There are no available hardware reservations for selected plan and facility"]}`

Since we can't do much if the request fails anyway, I felt it was more future proof to assume that the reservation might have been the issue and retry without one, instead of looking for the exact error code and message which might conceivably change in the future. In any case, I don't have a very strong preference here. I can make it more tied to the current behavior of the Packet API if that's preferred.

Reply:
Makes sense. If the error code is 503 when no reservation is done, it seems maybe better to retry ignoring it. It doesn't seem the best error code for this case :)